omnigibson/object_states/overlaid.py

```diff
@@ -20,13 +24,57 @@
 # Subsample cloth particle points to fit a convex hull for efficiency purpose
 m.N_POINTS_CONVEX_HULL = 1000
 
+# z-offset for sampling
+m.SAMPLING_Z_OFFSET = 0.01
+
 class Overlaid(KinematicsMixin, RelativeObjectState, BooleanState):
     @staticmethod
     def get_dependencies():
         return KinematicsMixin.get_dependencies() + RelativeObjectState.get_dependencies() + [Touching]
 
-    def _set_value(self, other, new_value, use_ray_casting_method=False):
-        raise NotImplementedError("Overlaid state currently does not support setting.")
+    def _set_value(self, other, new_value):
+        if not new_value:
+            raise NotImplementedError("Overlaid does not support set_value(False)")
+
+        if not (self.obj.prim_type == PrimType.CLOTH and other.prim_type == PrimType.RIGID):
+            raise ValueError("Overlaid state requires obj1 is cloth and obj2 is rigid.")
+
+        sampling_success = False
+        state = self._simulator.dump_state(serialized=False)
+
+        # Get the top center of the object below
+        aabb_low, aabb_hi = other.states[AABB].get_value()
+        top_center = (aabb_low + aabb_hi) / 2 + np.array([0., 0., (aabb_hi - aabb_low)[2] / 2.0])
+
+        # Reset the cloth
+        self.obj.root_link.reset()
+
+        # Add the half-extent of the object on top in the z-axis with additional offset
+        aabb_low, aabb_hi = self.obj.states[AABB].get_value()
+        pos = top_center + np.array([0., 0., (aabb_hi - aabb_low)[2] / 2.0 + m.SAMPLING_Z_OFFSET])
+
+        for _ in range(10):
+            # Use a random orientation in the z-axis
+            orn = T.euler2quat(np.array([0., 0., np.random.uniform(0, np.pi * 2)]))
+            self.obj.set_position_orientation(pos, orn)
+            self.obj.root_link.reset()
+            self.obj.keep_still()
+            # Let it fall for 0.2 second
+            for _ in range(int(0.2 / og.sim.get_physics_dt())):
+                og.sim.step_physics()
+                if len(self.obj.states[ContactBodies].get_value()) > 0:
+                    break
+
+            if self.get_value(other):
+                sampling_success = True
+                break
+            else:
+                if omnigibson.debug_sampling:
+                    print("Overlaid checking", sampling_success)
+                    embed()
```

Comment:
it would be nice to make this functionality an explicit util so we don't have IPython import polluting a bunch of modules

e.g. in ui_utils (or some other module):

```python
def debug_breakpoint(msg):
    logging.error(msg)
    embed()
```

Reply:
done! I also got rid of some of the debug breakpoint because I don't think they are useful there and they pollute the cleanness of the main object state codebase.